File: scripts/fetch-boja.py
# /// script
# requires-python = ">=3.10"
# dependencies = ["scrapling[fetchers]"]
# ///
"""fetch-boja.py — hijo de fetch-boja.mjs: consulta el buscador del BOJA.

El buscador de la sede del BOJA es Solr detrás de un GET plano (search.do):
la paginación expone fq/sort/start, así que esto es casi una API no documentada.
Se usa el Fetcher ligero de Scrapling (HTTP puro, sin navegador); si algún día
la Junta añade protección anti-bot, la escalera es StealthyFetcher/DynamicFetcher.

Este script NO decide qué es relevante ni qué es nuevo: trae la ventana completa
y emite JSON por stdout. El contrato, los invariantes y el filtro viven en el
padre Node. Los errores también viajan en el JSON ({ok:false,kind,msg}) — el
exit code del proceso lo interpreta el padre, no este script.

kinds de fallo: net (red/HTTP) · gone (search.do ya no existe) · shape (la
página responde pero su estructura ya no es la esperada — rediseño).
"""
import json
import logging
import re
import sys
from urllib.parse import urlencode

logging.getLogger("scrapling").setLevel(logging.ERROR)
from scrapling.fetchers import Fetcher  # noqa: E402

BASE = "https://www.juntadeandalucia.es/eboja/buscador/search.do"
PAGE_SIZE = 10  # tamaño fijo del listado del buscador


def fail(kind, msg):
    print(json.dumps({"ok": False, "kind": kind, "msg": msg}, ensure_ascii=False))
    sys.exit(0)


def get(params):
    try:
        return Fetcher.get(BASE + "?" + urlencode(params), timeout=30)
    except Exception as e:  # scrapling ya reintenta 3 veces por dentro
        fail("net", f"{type(e).__name__}: {e}")


def parse_total(page):
    """'1 a 10 de 23 resultados' → 23. None si no hay paginado (0 resultados… o rediseño)."""
    pag = page.css("div.paginado")
    if not pag:
        return None
    m = re.search(r"de\s+([\d.]+)\s+resultados", " ".join(pag[0].get_all_text().split()))
    return int(m.group(1).replace(".", "")) if m else None


def scan(q, start_date, max_pages):
    items, total, scanned = [], None, 0
    for pageno in range(max_pages):
        page = get({
            "eboja": "on",
            "q": q,
            "startDate": start_date,
            "ordenacion": "bojaDateNormalized",
            "sentido_ordenacion": "descendente",
            "start": pageno * PAGE_SIZE,
        })
        if page.status == 404:
            fail("gone", "search.do devuelve 404 — el buscador del BOJA ya no está donde estaba")
        if page.status != 200:
            fail("net", f"HTTP {page.status} en q={q!r}")
        lis = page.css("ul.listado_resultados > li")
        if pageno == 0:
            total = parse_total(page)
        if not lis:
            break
        for li in lis:
            a = li.css("a")
            items.append({
                "text": " ".join(li.get_all_text().split()),
                "link": (a[0].attrib.get("href") or "").strip() if a else None,
            })
        scanned += len(lis)
        if total is not None and scanned >= total:
            break
    # el paginado promete resultados que el listado no entrega: estructura rota
    if total and not items:
        fail("shape", f"el paginado dice {total} resultados pero el listado está vacío (q={q!r})")
    return {"q": q, "total": total if total is not None else scanned, "scanned": scanned, "items": items}


cfg = json.loads(sys.argv[1])
out = [scan(q, cfg["startDate"], cfg.get("maxPages", 5)) for q in cfg["queries"]]

# Cero en TODAS las consultas: ¿silencio real o buscador rediseñado? La página de
# cero resultados es idéntica a una rota (sin listado ni paginado), así que lo
# decide una consulta de control que siempre tiene resultados — mismo principio
# que la cuenta de control de fetch-x.mjs.
if all(r["scanned"] == 0 for r in out):
    control = scan(cfg.get("controlQuery", "decreto"), cfg["startDate"], 1)
    if control["scanned"] == 0:
        fail("shape", "todas las consultas Y la de control devuelven 0 items — se asume rediseño del buscador, no silencio real")

print(json.dumps({"ok": True, "queries": out}, ensure_ascii=False))

File: scripts/fetch-boja.mjs
#!/usr/bin/env node
/**
 * fetch-boja.mjs — vigía del BOJA (Boletín Oficial de la Junta de Andalucía).
 *
 * El expediente del incendio (ayudas, declaraciones, nombramientos) se publica
 * en el BOJA, que no tiene RSS útil. Su buscador sí es consultable: Solr detrás
 * de un GET plano. El fetch real lo hace scripts/fetch-boja.py (Scrapling, via
 * uv run) como proceso hijo; este script pone el contrato: invariantes, filtro,
 * dedupe y el parte por stdout. Hermano de fetch-news.mjs y fetch-x.mjs: emite
 * SOLO publicaciones no vistas y deja el detalle en data/boja/latest.json
 * (plano-máquina efímero, gitignored — alimenta el parte de /gather-updates).
 *
 * El buscador hace stemming ("Los Gallardos" ≈ apellido Gallardo) y su semántica
 * multi-término no es de fiar (medido: 'incendio forestal Almería'=0 resultados,
 * 'incendio'=16): consultas de UNA palabra o frase entrecomillada, ventana ancha
 * (fecha + orden descendente), y la relevancia la decide el filtro estricto de
 * filterPatterns aquí, en cliente. Un decreto de ayudas puede no nombrar al
 * municipio: por eso hay varias consultas y varios patrones, no una de cada.
 *
 * Invariantes como asserts (los de fetch-x):
 *   - sin enlace concreto al documento → descartado
 *   - sin fecha de boletín, o fecha FUTURA → descartado (invariante 4)
 *
 * Uso: node scripts/fetch-boja.mjs        (barrido real)
 *      node scripts/fetch-boja.mjs --dry  (fixture, sin red ni escritura)
 *
 * Exit: 0 ok · 3 forma/red rota (incluye rediseño del buscador) ·
 *       4 backend ausente (uv) · 5 capacidad upstream desaparecida (search.do ya no existe)
 */
import { readFileSync, writeFileSync, mkdirSync, existsSync } from 'node:fs';
import { spawnSync } from 'node:child_process';
import { fileURLToPath } from 'node:url';

const CFG = JSON.parse(readFileSync(new URL('../incident.config.json', import.meta.url), 'utf8'));
const B = CFG.boja ?? {};
const QUERIES = B.queries ?? [];
const START_DATE = B.startDate;
const PATTERNS = (B.filterPatterns ?? []).map(p => new RegExp(p, 'iu'));
const MAX_PAGES = B.maxPages ?? 5;
const BIN = B.bin ?? 'uv';
const PY = fileURLToPath(new URL('fetch-boja.py', import.meta.url));

const DRY = process.argv.includes('--dry');

const DIR = new URL('../data/boja/', import.meta.url);
if (!DRY) mkdirSync(DIR, { recursive: true });
const SEEN_PATH = new URL('seen.json', DIR);
const seen = existsSync(SEEN_PATH) ? new Set(JSON.parse(readFileSync(SEEN_PATH, 'utf8'))) : new Set();

const die = (code, msg) => { console.error(msg); process.exit(code); };

/* ---------- proceso hijo: uv run fetch-boja.py ---------- */
function runBackend() {
  const payload = { queries: QUERIES, startDate: START_DATE, maxPages: MAX_PAGES, controlQuery: B.controlQuery ?? 'decreto' };
  const res = spawnSync(BIN, ['run', '--quiet', PY, JSON.stringify(payload)], {
    env: { PATH: process.env.PATH, HOME: process.env.HOME, LANG: 'C.UTF-8' },
    encoding: 'utf8',
    timeout: 180000, // la primera ejecución resuelve el entorno Python (uv cachea después)
  });
  if (res.error?.code === 'ENOENT') die(4, `✗ BOJA: backend '${BIN}' no encontrado en PATH (instalar uv: https://docs.astral.sh/uv/)`);

  let out = null;
  try { out = JSON.parse(res.stdout); } catch { /* se decide abajo */ }

  if (out?.ok === false) {
    if (out.kind === 'gone') die(5, `✗ BOJA: capacidad upstream desaparecida — ${out.msg}\n  El barrido del BOJA queda SIN HACER: no confundir con "sin novedades".`);
    die(3, `✗ BOJA: ${out.kind === 'shape' ? 'estructura inesperada (¿rediseño del buscador?)' : 'fallo de red'} — ${out.msg}\n  El barrido del BOJA queda SIN HACER.`);
  }
  if (!Array.isArray(out?.queries)) {
    die(3, `✗ BOJA: salida ilegible del hijo Python (código ${res.status})\n${(res.stderr ?? '').trim().slice(-800)}`);
  }
  return out.queries;
}

/* ---------- validación: invariantes del proyecto, como asserts ----------
   El texto de cada item trae la ficha completa: "<título>. Organismo: <org>
   (Boletín número N de DD/MM/YYYY Sección: …)". Sin enlace o sin fecha de
   boletín, el item no sirve para el tier system y se descarta con motivo. */
function validate(raw, todayISO) {
  let link = (raw.link ?? '').trim();
  if (!/^https?:\/\//.test(link)) return { drop: 'sin enlace al documento' };
  link = link.replace(/^http:\/\/(www\.juntadeandalucia\.es)/, 'https://$1');

  const text = raw.text ?? '';
  const mBol = text.match(/\(?Bolet[íi]n n[úu]mero (\d+) de (\d{2})\/(\d{2})\/(\d{4})/);
  if (!mBol) return { drop: 'sin fecha de boletín' };
  const [, num, dd, mm, yyyy] = mBol;
  const date = `${yyyy}-${mm}-${dd}`;
  if (date > todayISO) return { drop: `fecha de boletín FUTURA (${date})` };

  const title = (text.split(/ Organismo:/)[0] ?? text).trim();
  const organismo = (text.match(/Organismo:\s*([^(]+)/)?.[1] ?? '').trim();
  return { ok: { url: link, title, organismo, boletin: num, publishedAt: date } };
}

/* ---------- ciclo ---------- */
if (!QUERIES.length || !START_DATE) die(3, '✗ BOJA: configuración incompleta (incident.config.json → boja.queries / boja.startDate)');

let results;
let todayISO;

if (DRY) {
  const fx = JSON.parse(readFileSync(new URL('fixtures/boja-search.json', import.meta.url), 'utf8'));
  results = fx.payload.queries;
  todayISO = fx._anchorToday;
  console.error(`· --dry: fixture boja-search.json · ancla ${todayISO} (sin red ni escritura)`);
} else {
  results = runBackend();
  todayISO = new Intl.DateTimeFormat('sv-SE', { timeZone: 'Europe/Madrid' }).format(new Date());
}

const dropped = [];
const fresh = [];
let scanned = 0;
let filteredOut = 0;
const dedup = new Set();

for (const r of results) {
  scanned += r.scanned;
  for (const raw of r.items) {
    const v = validate(raw, todayISO);
    if (v.drop) { dropped.push(v.drop); continue; }
    if (PATTERNS.length && !PATTERNS.some(p => p.test(raw.text))) { filteredOut++; continue; }
    if (dedup.has(v.ok.url) || seen.has(v.ok.url)) continue;
    dedup.add(v.ok.url);
    fresh.push(v.ok);
  }
}
fresh.sort((a, b) => b.publishedAt.localeCompare(a.publishedAt));

if (!DRY) {
  for (const it of fresh) seen.add(it.url);
  writeFileSync(SEEN_PATH, JSON.stringify([...seen].slice(-800), null, 1));
  writeFileSync(new URL('latest.json', DIR), JSON.stringify({ fetchedAt: new Date().toISOString(), items: fresh }, null, 1));
}

/* Techo declarado, nunca silencioso: la consulta va ordenada por fecha
   descendente, así que lo no escaneado es siempre lo más antiguo. */
for (const r of results) {
  if (r.total > r.scanned) console.error(`⚠ consulta ${r.q}: escaneados ${r.scanned} de ${r.total} (los más recientes primero; subir boja.maxPages si hace falta el fondo)`);
}
if (dropped.length) {
  const tally = dropped.reduce((a, d) => (a[d.replace(/\(.*\)/, '').trim()] = (a[d.replace(/\(.*\)/, '').trim()] ?? 0) + 1, a), {});
  console.error(`⚠ descartados ${dropped.length}: ${Object.entries(tally).map(([k, n]) => `${k} ×${n}`).join(', ')}`);
}

console.log(`· BOJA vía buscador oficial (${QUERIES.map(q => `'${q}'`).join(', ')}; desde ${START_DATE}). El buscador hace stemming y su multi-término no es fiable: la relevancia la decide el filtro en cliente, no él.`);

if (!fresh.length) {
  console.log(`✓ BOJA: sin publicaciones nuevas relevantes (${scanned} escaneadas, ${filteredOut} fuera de filtro)`);
} else {
  console.log(`✓ BOJA: ${fresh.length} publicaciones NUEVAS relevantes${DRY ? ' (--dry: nada escrito)' : ' (detalle en data/boja/latest.json)'}`);
  for (const it of fresh.slice(0, 20)) {
    const [y, m, d] = it.publishedAt.split('-');
    console.log(`- ${d}/${m} · BOJA ${it.boletin} · ${it.organismo || '¿organismo?'} · ${it.title.slice(0, 130)}${it.title.length > 130 ? '…' : ''}\n  ${it.url}`);
  }
  if (fresh.length > 20) console.log(`  … y ${fresh.length - 20} más en latest.json`);
}
